### PR TITLE
Try ISO-8859-1 as well as ascii string encoding

### PR DESCRIPTION
**Pull request by daehn**
# What's in this PR?

* There was a bug which prevented the completion handler of the link preview processor being called at all. For link with a non UTF-8 content type the completion handler wasn't invoked. We now try other encodings as well. This issue was discovered as the share extension will not dismiss until the link preview state is `.done`. 
* I'm still thinking about a better way to do this, e.g. checking the content-type header for charset information, as trying multiple encodings also is not the most robust way to handle this case.
* The link which blocked the share extension from dismissing was https://www.thalia.de/shop/home/show/ and the content uses an ISO-8859-1 charset.

### CODE REVIEW

**Comment by typfel** [comment]
Can we add a test case for this?

**Comment by daehn** [comment]
Surprised me as well, but the decoding fails using `utf8` and succeeds with `ascii` ¯\_(ツ)_/¯. Yeah I'll add a test for it. I am also adding another check independent of the decoding to ensure we always call the completion handler once the request completed.

**Comment by daehn** [comment]
`UTF-8` is indeed backwards compatibly with `ACII`, removed that case.

**Comment by daehn** [comment]
@typfel updated! I also added a check in `urlSession(session:task:didCompleteWithError:)` to ensure we call the completion handler if the task completed and we do not have any data.

File: WireLinkPreview/MetaStreamContainer.swift [hunk 1]
```diff
@@ -47,10 +47,15 @@ final class MetaStreamContainer {
     }
 
     private func updateReachedEndOfHead(withData data: Data) {
-        guard let string = String(data: data, encoding: String.Encoding.utf8)?.lowercased() else { return }
+        guard let string = String(data: data, encoding: .utf8)?.lowercased() else { return }
```
Comment (typfel) [inline WireLinkPreview/MetaStreamContainer.swift]: Why did you remove the use of `parseString` again?

File: WireLinkPreview/PreviewDownloader.swift [hunk 2]
```diff
@@ -69,8 +70,12 @@ final class PreviewDownloader: NSObject, URLSessionDataDelegate, PreviewDownload
     }
     
     func urlSession(_ session: URLSessionType, task: URLSessionDataTaskType, didCompleteWithError error: NSError?) {
+        guard let url = task.originalRequest?.url, let completion = completionByURL[url] else { return }
+        if let container = containerByTaskID[task.taskIdentifier], nil == container.stringContent && nil == error {
+            return completeAndCleanUp(completion, result: nil, url: url, taskIdentifier: task.taskIdentifier)
+        }
+
         guard let errorCode = error?.code , errorCode != URLError.cancelled.rawValue else { return }
```
Comment (typfel) [inline WireLinkPreview/PreviewDownloader.swift]: It's bit confusing with two seemingly identical `completeAndCleanUp` calls. Maybe this guard could be changed to allow `error == nil` and move to the top?

File: WireLinkPreviewTests/ImageDownloaderTests.swift [hunk 3]
```diff
@@ -126,7 +126,7 @@ class ImageDownloaderTests: XCTestCase {
         let completionExpectation = expectation(description: "It should call the completion handler")
         let url = URL(string: "www.example.com")!
         
-        let data = "test data".utf8Data
+        let data = "test data".data(using: .utf8)!
```
Comment (typfel) [inline WireLinkPreviewTests/ImageDownloaderTests.swift]: Stylistic change?

File: WireLinkPreviewTests/MetaStreamContainerTests.swift [hunk 4]
```diff
@@ -65,6 +56,38 @@ class MetaStreamContainerTests: XCTestCase {
     func testThatItSets_rechaedEndOfHead_WhenDataContainsHead_WithSpaces() {
         assertThatItUpdatesReachedEndOfHeadWhenItReceivedHead("</head >", shouldUpdate: false)
     }
+
+    func testThatItSets_rechaedEndOfHead_WhenDataContainsHead_Lowercase_Latin1() {
```
Comment (typfel) [inline WireLinkPreviewTests/MetaStreamContainerTests.swift]: typo: `rechaed` > reached

File: WireLinkPreview/MetaStreamContainer.swift (hunk 1 above)
Comment (daehn) [inline WireLinkPreview/MetaStreamContainer.swift]: Ups

File: WireLinkPreview/PreviewDownloader.swift (hunk 2 above)
Comment (daehn) [inline WireLinkPreview/PreviewDownloader.swift]: In the other case we cancel it ourselves and do not want to call the completion handler again.

File: WireLinkPreviewTests/ImageDownloaderTests.swift (hunk 3 above)
Comment (daehn) [inline WireLinkPreviewTests/ImageDownloaderTests.swift]: Removed the helper as it was only used in the 2 spots now.

**Comment by daehn** [comment]
@typfel updated!

**Comment by daehn** [comment]
@typfel please have a look again, I refactored the method in question a bit!
